File: Task#2/C++/Poclington.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <string>
#include <set>
#include <cmath>
#include <map>
#include <random>
#include <iomanip>

using namespace std;

int powmod(int a, int x, int mod){
    int res = 1;
    for (int i = 1; i <= x; i++){
        res = res * a;
        res = res % mod;
    }
    return res;
}

vector<int> sieveEratosthenes(int n) {
    vector<bool> is_prime(n + 1, true);
    vector<int> primes;
    
    is_prime[0] = is_prime[1] = false; // 0 и 1 не являются простыми
    
    for (int i = 2; i * i <= n; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= n; j += i) {
                is_prime[j] = false; // отмечаем все кратные i как составные
            }
        }
    }
    
    for (int i = 2; i <= n; i++) {
        if (is_prime[i]) {
            primes.push_back(i);
        }
    }
    
    return primes;
}

bool probability_Test(int num, int k){ // Миллер-Рабин
    if (num == 2 || num == 3){
        return true;
    }
    if (num < 2 || num % 2 == 0){
        return false;
    }
    int s = 0;
    int d = num-1;
    while (d % 2 == 0){
        d /= 2;
        s++;
    }
    for (int i = 0; i < k; i++){
        int a = rand() % (num - 4) + 2; // от 2 до num - 2
        int x = powmod(a, d, num);
        int y;
        for (int j = 0; j < s; j++){
            y = powmod(x, 2, num);
            if (y == 1 && x != 1 && x != (num - 1)) {
                return false;
            }
            x = y;
        }
        if (y != 1) {
            return false;
        }
    }
    return true;
}

int NOD(int a, int p) {     // проверка на взаимно простые числа
    while (p != 0) {        // поиск НОД (обобщенный алгоритм Евклида)
        int r = p;          // a и r переносят числа
        p = a % p;          // каждую итерацию после первой берем остаток от пред. делителя и остатка
        a = r;              // a%b = c >> b%c = g >> c%g = d
    }
    return a;
}

vector<int> canon(int n) {
    vector<int> prime = sieveEratosthenes(2000);
    vector<int> q;
    int N = n - 1;

    for (int i = 0; i < prime.size(); i++){
        int R = N;
        int k = 0;
        if (R % prime[i] == 0){
            while(R % prime[i] == 0){
                R /= prime[i];
                k += 1;
            }
            for (int j = 0; j < k; j++) {
                q.push_back(prime[i]);
            }
        }
        else {
            continue;
        }
    }

    int r, f;
    vector<int> F;

    while (true) {
        int r = 1;
        int f = 1;
        
        F.push_back(q[q.size()-1]);
        q.pop_back();

        for(int i = 0; i < F.size(); i++) {
            f = f * F[i];
        }

        for (int i = 0; i < q.size(); i++){
            r = r * q[i];
        }

        if (f > r && r % 2 == 0) {
            return F;
        }
        else {
            continue;
        }
    }
}

int test_Poclington(int n, int t) {
    vector<int> A;
    vector<int> check;

    if (n == 1 || n == 0) {
        return false;
    }

    vector<int> prime = sieveEratosthenes(8);
    for (int i = 0; i < prime.size(); i++) {
        if (n == prime[i]) {
            return true;
        }
    }

    for (int i = 0; i < t; i++){
        A.push_back(rand() % (n - 2) + 2);
    }

    for (int i = 0; i < A.size(); i++) {
        int res = powmod(A[i], n - 1, n);
        if (res != 1) {
            return false;
        }
    }

    vector<int> F = canon(n);

    for (int i = 0; i < A.size(); i++) {
        int temp_a = A[i]; 
        for (int i = 0; i < F.size(); i++) {
            int res = powmod(temp_a, (n - 1)/F[i], n);
            if (res == 1) {
                break;
            }
            else {
                check.push_back(res);
            }
        }
    }

    for (int i = 0; i < check.size(); i++) {
        if (check[i] == 1) {
            return false;
        }
        else {
            return true;
        }
    }

    return false;
}

int bitLen(int number){
    int bitCount = 0;
    while (number != 0){
        bitCount++;
        number /= 2;
    }
    return bitCount;
}

void printTable(const vector<int>& primeNums, const vector<bool>& primeTestAgree, const vector<int>& primeLostCount) {
    int n = primeNums.size();
    int colWidth = 12;

    auto printLine = [&](int count) {
        cout << "+";
        for (int i = 0; i <= count; i++) {
            cout << string(colWidth, '-') << "+";
        }
        cout << endl;
    };

    printLine(n);
    cout << "| " << setw(colWidth + 1) << left << "№";
    for (int i = 0; i < n; i++) {
        cout << "| " << setw(colWidth - 1) << left << i + 1;
    }
    cout << "|" << endl;

    printLine(n);

    cout << "| " << setw(colWidth - 1) << left << "P";
    for (int num : primeNums) {
        cout << "| " << setw(colWidth - 1) << left << num;
    }
    cout << "|" << endl;
    printLine(n);

    cout << "| " << setw(colWidth + 3) << left << "Тест";
    for (bool result : primeTestAgree) {
        cout << "| " << setw(colWidth - 1) << left << (result ? "+" : "-");
    }
    cout << "|" << endl;
    printLine(n);

    cout << "| " << setw(colWidth - 1) << left << "K";
    for (int k : primeLostCount) {
        cout << "| " << setw(colWidth - 1) << left << k;
    }
    cout << "|" << endl;
    printLine(n);

}

int main() {

    int t = 5;
    int test_k = 3;
    int m;
    int r;
    bool isPrime;
    int k;
    vector<int> primeNums;
    vector<int> primeLostCount;
    vector<bool> primeTestAgree;
    for (int g = 0; g < 10; g++){
        isPrime = false;
        k = 0;
        while(!isPrime){
            random_device rd;
            mt19937 gen(rd());

            // генерация случайного числа в диапазоне (6, 16)
            uniform_int_distribution<> bitDist(6, 16);
            int bit = bitDist(gen);

            int F = 1;
            vector<int> sieve = sieveEratosthenes(500);

            // генерация F
            uniform_int_distribution<> sieveDist(0, sieve.size() - 1);
            while (F < pow(2, bit / 2) || F > pow(2, bit / 2 + 1)) {
                int temp = sieveDist(gen);
                F = F * sieve[temp];

                if (F > pow(2, bit / 2 + 1)) {
                    F = 1;
                }
            }

            // генерация n
            int n = 0;
            uniform_int_distribution<> nDist(pow(2, bit / 2 - 1), pow(2, bit / 2) - 1);
            for (int r = nDist(gen); r < pow(2, bit / 2); ++r) {
                if (r % 2 == 0) {
                    n = r * F;
                    if (bitLen(n) == bit) {
                        n = n + 1;
                        break;
                    }
                }
            }
            int proverka = 0;
            for (int i = 0; i < primeNums.size(); i++) {
                if (primeNums[i] == n) {
                    proverka += 1;
                }
            }

            if (proverka != 0) {
                continue;
            }

            if (n != 1 && test_Poclington(n, t)) {
                primeNums.push_back(n); 
                isPrime = true;
                if (probability_Test(n, test_k)){ 
                    primeTestAgree.push_back(true);
                }
                else{
                    primeTestAgree.push_back(false);
                }
            }
            else {
                if (probability_Test(n, test_k)){ 
                    k++;
                }
            }
        }
        primeLostCount.push_back(k);
    }

    cout << setw(95) << "Тест Поклингтона" << endl;
    printTable(primeNums, primeTestAgree, primeLostCount);
}

File: Task#2/Python/Poclington.py
import random
import math
from typing import List, Tuple

def powmod(a: int, x: int, mod: int) -> int:
    res = 1
    for _ in range(x):
        res = res * a
        res = res % mod
    return res

def sieveEratosthenes(n: int) -> List[int]:
    is_prime = [True] * (n + 1)
    primes = []
    
    is_prime[0] = is_prime[1] = False  # 0 и 1 не являются простыми
    
    for i in range(2, int(math.sqrt(n)) + 1):
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False  # отмечаем все кратные i как составные
    
    for i in range(2, n + 1):
        if is_prime[i]:
            primes.append(i)
    
    return primes

def probability_Test(num: int, k: int) -> bool:  # Миллер-Рабин
    if num == 2 or num == 3:
        return True
    if num < 2 or num % 2 == 0:
        return False
    s = 0
    d = num - 1
    while d % 2 == 0:
        d //= 2
        s += 1
    for _ in range(k):
        a = random.randint(2, num - 2)  # от 2 до num - 2
        x = powmod(a, d, num)
        y = 0
        for _ in range(s):
            y = powmod(x, 2, num)
            if y == 1 and x != 1 and x != (num - 1):
                return False
            x = y
        if y != 1:
            return False
    return True

def NOD(a: int, p: int) -> int:  # проверка на взаимно простые числа
    while p != 0:  # поиск НОД (обобщенный алгоритм Евклида)
        r = p  # a и r переносят числа
        p = a % p  # каждую итерацию после первой берем остаток от пред. делителя и остатка
        a = r  # a%b = c >> b%c = g >> c%g = d
    return a

def canon(n: int) -> List[int]:
    prime = sieveEratosthenes(2000)
    q = []
    N = n - 1

    for p in prime:
        R = N
        k = 0
        if R % p == 0:
            while R % p == 0:
                R //= p
                k += 1
            for _ in range(k):
                q.append(p)
        else:
            continue

    F = []

    while True:
        r = 1
        f = 1
        
        F.append(q[-1])
        q.pop()

        for factor in F:
            f *= factor

        for factor in q:
            r *= factor

        if f > r and r % 2 == 0:
            return F
        else:
            continue

def test_Poclington(n: int, t: int) -> bool:
    A = []
    check = []

    if n == 1 or n == 0:
        return False

    prime = sieveEratosthenes(8)
    if n in prime:
        return True

    for _ in range(t):
        A.append(random.randint(2, n - 2))

    for a in A:
        res = powmod(a, n - 1, n)
        if res != 1:
            return False

    F = canon(n)

    for a in A:
        temp_a = a
        for f in F:
            res = powmod(temp_a, (n - 1) // f, n)
            if res == 1:
                break
            else:
                check.append(res)

    for c in check:
        if c == 1:
            return False
        else:
            return True

    return False

def bitLen(number: int) -> int:
    bitCount = 0
    while number != 0:
        bitCount += 1
        number //= 2
    return bitCount

def printTable(primeNums: List[int], primeTestAgree: List[bool], primeLostCount: List[int]):
    n = len(primeNums)
    colWidth = 12

    def printLine(count: int):
        print("+", end="")
        for _ in range(count + 1):
            print("-" * colWidth + "+", end="")
        print()

    printLine(n)
    print("| {:<{}}".format("№", colWidth + 1), end="")
    for i in range(n):
        print("| {:<{}}".format(i + 1, colWidth - 1), end="")
    print("|")

    printLine(n)

    print("| {:<{}}".format("P", colWidth - 1), end="")
    for num in primeNums:
        print("| {:<{}}".format(num, colWidth - 1), end="")
    print("|")
    printLine(n)

    print("| {:<{}}".format("Тест", colWidth + 3), end="")
    for result in primeTestAgree:
        print("| {:<{}}".format("+" if result else "-", colWidth - 1), end="")
    print("|")
    printLine(n)

    print("| {:<{}}".format("K", colWidth - 1), end="")
    for k in primeLostCount:
        print("| {:<{}}".format(k, colWidth - 1), end="")
    print("|")
    printLine(n)

def main():
    t = 5
    test_k = 3
    primeNums = []
    primeLostCount = []
    primeTestAgree = []

    for _ in range(10):
        isPrime = False
        k = 0
        while not isPrime:
            bit = random.randint(6, 16)

            F = 1
            sieve = sieveEratosthenes(500)

            while F < 2 ** (bit // 2) or F > 2 ** (bit // 2 + 1):
                temp = random.choice(sieve)
                F *= temp

                if F > 2 ** (bit // 2 + 1):
                    F = 1

            n = 0
            r = random.randint(2 ** (bit // 2 - 1), 2 ** (bit // 2) - 1)
            if r % 2 == 0:
                n = r * F
                if bitLen(n) == bit:
                    n += 1

            if n in primeNums:
                continue

            if n != 1 and test_Poclington(n, t):
                primeNums.append(n)
                isPrime = True
                if probability_Test(n, test_k):
                    primeTestAgree.append(True)
                else:
                    primeTestAgree.append(False)
            else:
                if probability_Test(n, test_k):
                    k += 1

        primeLostCount.append(k)

    print("{:^95}".format("Тест Поклингтона"))
    printTable(primeNums, primeTestAgree, primeLostCount)

if __name__ == "__main__":
    main()